src/Xamarin.Android.Build.Tasks/Tasks/CompileNativeAssembly.cs

```diff
@@ -0,0 +1,211 @@
+				CreateNoWindow = true,
+				WindowStyle = ProcessWindowStyle.Hidden,
+				WorkingDirectory = WorkingDirectory,
+			};
+
+			string assemblerName = Path.GetFileName (config.AssemblerPath);
+			LogDebugMessage ($"[Native Assembler] {psi.FileName} {psi.Arguments}");
+			using (var proc = new Process ()) {
+				proc.OutputDataReceived += (s, e) => {
+					if (e.Data != null)
+						OnOutputData (assemblerName, s, e);
+					else
+						stdout_completed.Set ();
+				};
+
+				proc.ErrorDataReceived += (s, e) => {
+					if (e.Data != null)
+						OnErrorData (assemblerName, s, e);
+					else
+						stderr_completed.Set ();
+				};
+
+				proc.StartInfo = psi;
+				proc.Start ();
+				proc.BeginOutputReadLine ();
+				proc.BeginErrorReadLine ();
+				Token.Register (() => { try { proc.Kill (); } catch (Exception) { } });
+				proc.WaitForExit ();
+
+				if (psi.RedirectStandardError)
+					stderr_completed.WaitOne (TimeSpan.FromSeconds (30));
+
+				if (psi.RedirectStandardOutput)
+					stdout_completed.WaitOne (TimeSpan.FromSeconds (30));
+
+				return proc.ExitCode == 0;
+			}
+		}
+
+		IEnumerable<Config> GetAssemblerConfigs ()
+		{
+			string sdkBinDirectory = MonoAndroidHelper.GetOSBinPath ();
+			foreach (ITaskItem item in Sources) {
+				string abi = item.GetMetadata ("abi")?.ToLowerInvariant ();
+				string prefix = String.Empty;
+				AndroidTargetArch arch;
+
+				switch (abi) {
+					case "armeabi-v7a":
+						prefix = Path.Combine (sdkBinDirectory, "arm-linux-androideabi");
+						arch = AndroidTargetArch.Arm;
+						break;
+
+					case "arm64":
+					case "arm64-v8a":
+					case "aarch64":
+						prefix = Path.Combine (sdkBinDirectory, "aarch64-linux-android");
+						arch = AndroidTargetArch.Arm64;
+						break;
+
+					case "x86":
+						prefix = Path.Combine (sdkBinDirectory, "i686-linux-android");
+						arch = AndroidTargetArch.X86;
+						break;
+
+					case "x86_64":
+						prefix = Path.Combine (sdkBinDirectory, "x86_64-linux-android");
+						arch = AndroidTargetArch.X86_64;
+						break;
+
+					default:
+						throw new NotSupportedException ($"Unsupported Android target architecture ABI: {abi}");
+				}
+
+				// We don't need the directory since our WorkingDirectory is (and must be) where all the
+				// sources are (because of the typemap.inc file being included by the other sources with
+				// a relative path of `.`)
+				string sourceFile = Path.GetFileName (item.ItemSpec);
+				var assemblerOptions = new List<string> {
+					"--warn",
+					"-o",
+					QuoteFileName (sourceFile.Replace (".s", ".o"))
+				};
+
+				if (DebugBuild)
+					assemblerOptions.Add ("-g");
+
+				assemblerOptions.Add (QuoteFileName (sourceFile));
+
+				yield return new Config {
+					InputSource = item.ItemSpec,
+					AssemblerPath = $"{prefix}-as",
+					AssemblerOptions = String.Join (" ", assemblerOptions),
+				};
+			}
+		}
+
+		void OnOutputData (string assemblerName, object sender, DataReceivedEventArgs e)
+		{
+			if (e.Data != null)
```

Comment:
Do we need to care about non-`null` data?  Worst case, we just wind up with an "empty" line dumped to diag build output, so why check?